blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +267,225 @@ where
+            return Ok(Some((tipset, receipt)));
+        }
+        self.search_back_for_message(&tipset, message)
+    }
+    /// returns a message receipt from a given tipset and message cid
+    pub fn get_receipt(&self, tipset: &Tipset, msg: &Cid) -> Result<MessageReceipt, Error> {
+        let m = chain::get_chain_message(self.get_block_store_ref(), msg)
+            .map_err(|e| Error::Other(e.to_string()))?;
+        let message_receipt = self.tipset_executed_message(tipset, msg, &*m)?;
+
+        if let Some(receipt) = message_receipt {
+            return Ok(receipt);
+        }
+
+        let maybe_tuple = self.search_back_for_message(tipset, &*m)?;
+        let message_receipt = maybe_tuple
+            .ok_or_else(|| {
+                Error::Other("Could not get receipt from search back message".to_string())
+            })?
+            .1;
+        Ok(message_receipt)
+    }
+
+    fn tipset_executed_message(
+        &self,
+        tipset: &Tipset,
+        cid: &Cid,
+        message: &dyn Message,
+    ) -> Result<Option<MessageReceipt>, Error> {
+        if tipset.epoch() == 0 {
+            return Ok(None);
+        }
+        let tipset = chain::tipset_from_keys(self.get_block_store_ref(), tipset.parents())
+            .map_err(|err| Error::Other(err.to_string()))?;
+        let messages = chain::messages_for_tipset(self.get_block_store_ref(), &tipset)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        messages
+            .iter()
+            .enumerate()
+            .rev()
+            .filter(|(_, s)| s.from() == message.from())
+            .filter_map(|(index,s)| {
+                if s.sequence() == message.sequence() {
+                    if s.to_cid().map(|s| &s == cid).unwrap_or_default() {
+                        return Some(
+                            chain::get_parent_reciept(
+                                self.get_block_store_ref(),
+                                tipset.blocks().first().unwrap(),
+                                index as u64,
+                            )
+                            .map_err(|err| {
+                                Error::Other(err.to_string())
+                            }),
+                        );
+                    }
+                    if let Ok(m_cid) = s.to_cid()
+                    {
+                        let error_msg = format!("found message with equal nonce as the one we are looking for (F:{:} n {:}, TS: {:} n{:})",cid,message.sequence(),m_cid,s.sequence());
+                        return Some(Err(Error::Other(error_msg)))
+                    }
+                    let error_msg =format!("found message with equal nonce as the one we are looking for (F:{:} n {:}, TS: `Error Converting message to Cid` n{:})",cid,message.sequence(),s.sequence());
+                    return Some(Err(Error::Other(error_msg)))
+                }
+                if s.sequence() < message.sequence() {
+                    return Some(Ok(None));
```

Comment:
Are you sure this isn't:
```suggestion
                    return None;
```
Seems like that is what is intended here, otherwise you will keep checking messages even though the sequence cannot be equal to anymore

Reply:
Since it is a filter_map if it hits this comparison it should propagate this further. There needs to be a condition that calls for https://github.com/filecoin-project/lotus/blob/c4de2373fd13c909876cf459e59ead7749b9939d/chain/stmgr/stmgr.go#L698 if a None is returned the filter_map will go to the next one

Reply:
oh gotcha okay, yeah that logic block was confusing to me on first read